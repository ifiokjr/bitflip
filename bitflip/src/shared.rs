pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
