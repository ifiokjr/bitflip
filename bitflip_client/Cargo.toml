[package]
name = "bitflip_client"
version = { workspace = true }
authors = { workspace = true }
edition = { workspace = true }
homepage = { workspace = true }
license = { workspace = true }
repository = { workspace = true }
rust-version = { workspace = true }
description = "The rust based client for interacting with the bitflip solana program `bitflip`."

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_client"

[dependencies]
anchor-lang = { workspace = true }
anchor-spl = { workspace = true }
async-trait = { workspace = true }
bitflip_program = { workspace = true, features = ["client"] }
derive_more = { workspace = true, features = ["debug", "deref", "deref_mut"] }
futures = { workspace = true }
log = { workspace = true }
paste = { workspace = true }
solana-program = { workspace = true }
solana-sdk = { workspace = true }
typed-builder = { workspace = true }
wallet_standard = { workspace = true, features = ["solana"] }
wasm_client_anchor = { workspace = true }
wasm_client_solana = { workspace = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
solana-program-test = { workspace = true }
spl-associated-token-account = { workspace = true }
spl-transfer-hook-interface = { workspace = true }
test-log = { workspace = true, features = ["log", "trace"] }
test_utils = { workspace = true }
test_utils_solana = { workspace = true, features = ["test_validator"] }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
wallet_standard_wallets = { workspace = true }

[features]
ssr = ["wasm_client_solana/ssr"]
js = [
	"wasm_client_solana/js",
	"wasm_client_anchor/js",
	"bitflip_program/no-entrypoint",
]
