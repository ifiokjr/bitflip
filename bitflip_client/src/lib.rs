pub use crate::client::*;
pub use crate::pda::*;

pub mod client;
mod pda;
pub mod prelude {
	pub use wasm_client_anchor::prelude::*;
}
