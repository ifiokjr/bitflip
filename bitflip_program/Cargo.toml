[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
