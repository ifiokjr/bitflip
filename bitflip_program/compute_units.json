{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
