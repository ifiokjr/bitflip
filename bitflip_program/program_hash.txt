ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
