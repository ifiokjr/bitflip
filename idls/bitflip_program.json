{
	"address": "5AuNvfV9Xi9gskJpW2qQJndQkFcwbWNV6fjaf2VvuEcM",
	"metadata": {
		"name": "bitflip_program",
		"version": "0.0.0",
		"spec": "0.1.0",
		"description": "Created with Anchor"
	},
	"instructions": [
		{
			"name": "initialize_bits_data_section",
			"docs": [
				"This will initialize a single section of the bits data. It should be",
				"called 16 times with the correct index of the data chunch."
			],
			"discriminator": [
				171,
				216,
				141,
				120,
				246,
				225,
				197,
				148
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data for the full bits state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							}
						]
					}
				},
				{
					"name": "bits_data_section",
					"docs": [
						"This is a section of the bits data being initialized."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115,
									95,
									115,
									101,
									99,
									116,
									105,
									111,
									110
								]
							},
							{
								"kind": "arg",
								"path": "section"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "section",
					"type": "u8"
				}
			]
		},
		{
			"name": "initialize_bits_meta",
			"docs": [
				"This will initialize the meta state for the bits."
			],
			"discriminator": [
				202,
				134,
				244,
				179,
				235,
				49,
				7,
				137
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"Can't initialize the full state at one time, must incrementally add the",
						"state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": []
		},
		{
			"name": "initialize_config",
			"docs": [
				"Initialize the configuration for the program."
			],
			"discriminator": [
				208,
				127,
				21,
				1,
				194,
				190,
				196,
				70
			],
			"accounts": [
				{
					"name": "config",
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "admin",
					"docs": [
						"The hard-coded account that is used to initialize the program config",
						"once."
					],
					"writable": true,
					"signer": true,
					"address": "4z5X2suocz9szaQnSshj2AW8tuLgUVmYUxiW9hhPaRHs"
				},
				{
					"name": "treasury",
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeConfigProps"
						}
					}
				}
			]
		},
		{
			"name": "initialize_token",
			"docs": [
				"Initialize the token account. This must be called before the first game",
				"starts to generate the reward token."
			],
			"discriminator": [
				38,
				209,
				150,
				50,
				190,
				117,
				16,
				54
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The program authority which must be a signer to create this token."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "mint",
					"writable": true
				},
				{
					"name": "treasury",
					"writable": true
				},
				{
					"name": "treasury_token_account",
					"writable": true
				},
				{
					"name": "associated_token_program"
				},
				{
					"name": "token_program"
				},
				{
					"name": "system_program"
				},
				{
					"name": "bitflip_program",
					"docs": [
						"The program that is for signing."
					],
					"address": "5AuNvfV9Xi9gskJpW2qQJndQkFcwbWNV6fjaf2VvuEcM"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeTokenProps"
						}
					}
				}
			]
		},
		{
			"name": "initialize_token_inner",
			"docs": [
				"INNER: This method can only be called by [`initialize_token`]. It uses",
				"the `treasury` signer as the authority for the mint account."
			],
			"discriminator": [
				112,
				121,
				158,
				27,
				165,
				27,
				94,
				26
			],
			"accounts": [
				{
					"name": "authority",
					"docs": [
						"CHECKED: checked in [`InitializeToken`] outer call."
					],
					"writable": true,
					"signer": true
				},
				{
					"name": "mint",
					"docs": [
						"The token mint account."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									109,
									105,
									110,
									116
								]
							}
						]
					}
				},
				{
					"name": "treasury",
					"docs": [
						"The treasury account."
					],
					"writable": true,
					"signer": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "treasury_token_account",
					"docs": [
						"The associated token account for the treasury which will hold the minted",
						"tokens."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "treasury"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "associated_token_program",
					"docs": [
						"The program for associated tokens"
					],
					"address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
				},
				{
					"name": "token_program",
					"docs": [
						"The token program."
					],
					"address": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
				},
				{
					"name": "system_program",
					"docs": [
						"Needed in case a reallocation is required for the project memory."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeTokenInnerProps"
						}
					}
				}
			],
			"returns": "u8"
		},
		{
			"name": "set_bits",
			"docs": [
				"Flip bits based on the provided props."
			],
			"discriminator": [
				67,
				62,
				24,
				241,
				239,
				200,
				236,
				192
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "mint",
					"docs": [
						"The token mint account."
					],
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									109,
									105,
									110,
									116
								]
							}
						]
					}
				},
				{
					"name": "treasury",
					"docs": [
						"The treasury account which will transfer the spl tokens to the player."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "treasury_token_account",
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "treasury"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data for the full bits state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							}
						]
					}
				},
				{
					"name": "bits_data_section",
					"docs": [
						"The data for this section of the bit canvas."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115,
									95,
									115,
									101,
									99,
									116,
									105,
									111,
									110
								]
							},
							{
								"kind": "arg",
								"path": "props.section"
							}
						]
					}
				},
				{
					"name": "player",
					"docs": [
						"The player of the bit games"
					],
					"writable": true,
					"signer": true
				},
				{
					"name": "player_token_account",
					"docs": [
						"The associated token account for the main authority."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "player"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "associated_token_program",
					"docs": [
						"The program for associated tokens"
					],
					"address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
				},
				{
					"name": "token_program",
					"docs": [
						"The token program."
					],
					"address": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
				},
				{
					"name": "system_program",
					"docs": [
						"Needed for cpi payment instructions instructions."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "SetBitsProps"
						}
					}
				}
			]
		},
		{
			"name": "start_bits_session",
			"docs": [
				"Start the bits session with and set the flipped bits to the correct",
				"number. This can only be called once 16 sections have been initialized."
			],
			"discriminator": [
				86,
				86,
				164,
				215,
				248,
				251,
				126,
				195
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data account for the game."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "flipped_bits",
					"type": "u32"
				}
			]
		}
	],
	"accounts": [
		{
			"name": "BitsDataSectionState",
			"discriminator": [
				140,
				124,
				237,
				202,
				139,
				138,
				99,
				116
			]
		},
		{
			"name": "BitsMetaState",
			"discriminator": [
				93,
				244,
				217,
				115,
				235,
				203,
				72,
				85
			]
		},
		{
			"name": "ConfigState",
			"discriminator": [
				193,
				77,
				160,
				128,
				208,
				254,
				180,
				135
			]
		}
	],
	"errors": [
		{
			"code": 6000,
			"name": "BitsIncreaseSpaceInvalid",
			"msg": "The space is already fully initialized"
		},
		{
			"code": 6001,
			"name": "BitsUnchanged",
			"msg": "No update recorded"
		},
		{
			"code": 6002,
			"name": "InvalidAccount",
			"msg": "The provided account was invalid"
		},
		{
			"code": 6003,
			"name": "InvalidBitsLength",
			"msg": "The bits array is an invalid length"
		},
		{
			"code": 6004,
			"name": "InvalidBitsDataSection",
			"msg": "Invalid bit data section requested"
		},
		{
			"code": 6005,
			"name": "InvalidBitsDataSectionIndex",
			"msg": "Invalid bits data section index requested"
		},
		{
			"code": 6006,
			"name": "InvalidBitsDataSectionLength",
			"msg": "Invalid bits data section array length"
		},
		{
			"code": 6007,
			"name": "InvalidBitsDataSectionOrder",
			"msg": "Data sections initialized out of order"
		},
		{
			"code": 6008,
			"name": "InvalidFlippedBits",
			"msg": "An invalid number of flipped bits was provided"
		},
		{
			"code": 6009,
			"name": "Invalid256BitsDataSectionIndex",
			"msg": "The data section index must be a multiple of 16"
		},
		{
			"code": 6010,
			"name": "InvalidBitOffset",
			"msg": "The bit offset is invalid and must be less than 16"
		},
		{
			"code": 6011,
			"name": "NotRunning",
			"msg": "The current `BitsMetaState` is not running"
		},
		{
			"code": 6012,
			"name": "UnauthorizedAdmin",
			"msg": "The admin used was incorrect"
		},
		{
			"code": 6013,
			"name": "AllSectionsInitialized",
			"msg": "All bit data sections have already been initialized"
		},
		{
			"code": 6014,
			"name": "TokenNotInitialized",
			"msg": "The token is not yet initialized"
		}
	],
	"types": [
		{
			"name": "BitsDataSectionState",
			"docs": [
				"The data for each section of the the data. The total data is split into 16",
				"sections and this is one of the sections."
			],
			"serialization": "bytemuckunsafe",
			"repr": {
				"kind": "c"
			},
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "data",
						"docs": [
							"The state of the bits that are represented as flippable bits on the",
							"frontend."
						],
						"type": {
							"array": [
								"u16",
								4096
							]
						}
					}
				]
			}
		},
		{
			"name": "BitsMetaState",
			"docs": [
				"Adding [`BitState::on`] to [`BitState::off`] should always equal `1_000_000`"
			],
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "start_time",
						"docs": [
							"The start time. If 0 then it hasn't started yet. Using an `Option` here",
							"would waste an extra byte."
						],
						"type": "i64"
					},
					{
						"name": "flips",
						"docs": [
							"The number of bit flips that have occurred."
						],
						"type": "u64"
					},
					{
						"name": "on",
						"docs": [
							"The number of bits that are on."
						],
						"type": "u32"
					},
					{
						"name": "off",
						"docs": [
							"The number of bits that are off."
						],
						"type": "u32"
					},
					{
						"name": "index",
						"docs": [
							"The index of this currently played game."
						],
						"type": "u8"
					},
					{
						"name": "bump",
						"docs": [
							"The bump for this account."
						],
						"type": "u8"
					},
					{
						"name": "section_bumps",
						"docs": [
							"The bumps for each initialized section"
						],
						"type": "bytes"
					}
				]
			}
		},
		{
			"name": "ConfigState",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "authority",
						"docs": [
							"The authority which can update this config."
						],
						"type": "pubkey"
					},
					{
						"name": "lamports_per_bit",
						"docs": [
							"The number of lamports per bit change."
						],
						"type": "u64"
					},
					{
						"name": "bump",
						"docs": [
							"Store the bump to save compute units."
						],
						"type": "u8"
					},
					{
						"name": "treasury_bump",
						"docs": [
							"The treasury account bump where fees are sent and where the minted",
							"tokens are transferred."
						],
						"type": "u8"
					},
					{
						"name": "mint_bump",
						"docs": [
							"The mint account bump."
						],
						"type": "u8"
					},
					{
						"name": "bits_index",
						"docs": [
							"There will be a maximum of 4 games."
						],
						"type": "u8"
					}
				]
			}
		},
		{
			"name": "InitializeConfigProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "authority",
						"docs": [
							"The authority is the solana account which will have rights to update",
							"this configuration."
						],
						"type": "pubkey"
					}
				]
			}
		},
		{
			"name": "InitializeTokenInnerProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "name",
						"type": "string"
					},
					{
						"name": "symbol",
						"type": "string"
					},
					{
						"name": "uri",
						"type": "string"
					},
					{
						"name": "treasury_bump",
						"type": "u8"
					}
				]
			}
		},
		{
			"name": "InitializeTokenProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "name",
						"type": "string"
					},
					{
						"name": "symbol",
						"type": "string"
					},
					{
						"name": "uri",
						"type": "string"
					}
				]
			}
		},
		{
			"name": "SetBitsProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "section",
						"docs": [
							"The data section being updated."
						],
						"type": "u8"
					},
					{
						"name": "index",
						"docs": [
							"The index of the bit being set."
						],
						"type": "u16"
					},
					{
						"name": "variant",
						"docs": [
							"The new bit values."
						],
						"type": {
							"defined": {
								"name": "SetBitsVariant"
							}
						}
					}
				]
			}
		},
		{
			"name": "SetBitsVariant",
			"type": {
				"kind": "enum",
				"variants": [
					{
						"name": "On",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Off",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Bit16",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Bits256",
						"fields": [
							{
								"vec": "u16"
							}
						]
					}
				]
			}
		}
	]
}
