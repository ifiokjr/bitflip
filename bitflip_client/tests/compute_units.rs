use assert2::check;
use bitflip_client::BitflipProgramClient;
use bitflip_client::get_pda_config;
use bitflip_client::get_pda_mint;
use bitflip_client::get_pda_treasury;
use bitflip_program::ConfigState;
use bitflip_program::ID_CONST;
use bitflip_program::InitializeConfigProps;
use bitflip_program::accounts::InitializeConfig;
use shared::TestBitflipProgramClient;
use shared::create_admin_keypair;
use shared::create_authority_keypair;
use shared::create_runner;
use shared::create_runner_with_accounts;
use shared::create_treasury_keypair;
use solana_sdk::account::AccountSharedData;
use solana_sdk::signer::Signer;
use solana_sdk::system_program;
use test_log::test;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;
use wallet_standard_wallets::MemoryWallet;

mod shared;

#[test(tokio::test)]
async fn initialize_config_test() -> anyhow::Result<()> {
	let admin_keypair = create_admin_keypair();
	let admin = admin_keypair.pubkey();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let system_program = system_program::id();
	let (config, config_bump) = get_pda_config();
	let (treasury, treasury_bump) = get_pda_treasury();
	let mut runner = create_runner().await;
	let rpc = runner.rpc();
	let wallet = MemoryWallet::new(rpc.clone(), &[admin_keypair.insecure_clone()]);
	let bitflip_program_client: TestBitflipProgramClient = BitflipProgramClient::builder()
		.rpc(rpc.clone())
		.wallet(wallet)
		.build()
		.into();

	let initialize_config_request = bitflip_program_client
		.initialize_config()
		.args(InitializeConfigProps { authority })
		.accounts(InitializeConfig {
			config,
			admin,
			treasury,
			system_program,
		})
		.signers(vec![&admin_keypair])
		.build();

	let simulation = initialize_config_request
		.sign_and_simulate_transaction()
		.await?;

	log::info!("units_consumed: {simulation:#?}");
	check!(simulation.value.units_consumed.unwrap() < 75_000);

	let signature = initialize_config_request
		.sign_and_send_transaction()
		.await?;

	let config_state_account = bitflip_program_client
		.account::<ConfigState>(&config)
		.await?;

	insta::assert_yaml_snapshot!(config_state_account);

	Ok(())
}

fn create_config_state() -> AccountSharedData {
	let authority = create_authority_keypair().pubkey();
	let (_, bump) = get_pda_config();
	let (_, treasury_bump) = get_pda_treasury();
	let state = InitializeConfigProps { authority }.into_launchpad_state(bump, treasury_bump);

	AccountSharedData::from_anchor_data(state, ID_CONST)
}
