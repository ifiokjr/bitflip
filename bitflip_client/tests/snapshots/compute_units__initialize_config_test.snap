---
source: bitflip_client/tests/compute_units.rs
expression: config_state_account
---
authority:
  - 204
  - 41
  - 109
  - 182
  - 252
  - 170
  - 242
  - 95
  - 90
  - 190
  - 103
  - 20
  - 66
  - 112
  - 0
  - 95
  - 11
  - 33
  - 201
  - 29
  - 129
  - 252
  - 77
  - 105
  - 38
  - 75
  - 95
  - 65
  - 23
  - 191
  - 53
  - 214
lamportsPerBit: 100000
bump: 253
treasuryBump: 252
mintBump: ~
bitsIndex: 0
